api/base/parsers.py

```diff
@@ -15,6 +17,29 @@ class JSONAPIParser(JSONParser):
     media_type = 'application/vnd.api+json'
     renderer_class = JSONAPIRenderer
 
+    def flatten_relationships(self, relationships):
+        """
+        Flattens relationships dictionary which has information needed to create related resource objects.
+
+        Validates that formatting of relationships dictionary is correct.
+        """
+        if not isinstance(relationships, dict):
+                raise ParseError()
+        related_resource = relationships.keys()[0]
+        if not isinstance(relationships[related_resource], dict) or related_resource == 'data':
+            raise ParseError()
+        data = relationships[related_resource].get('data')
+
+        if not data:
+            raise JSONAPIException(source={'pointer': '/relationships/<related_resource_name>/data'}, detail=NO_DATA_ERROR)
```

Comment:
Discussed with @samchrisinger who suggests I could make a subclass JSONAPIRelationshipError but is not tied to the idea.  In my opinion, the specific sources used when validating the `relationships` dictionary are highly specific and will not occur again as the dictionary is flattened at this point.  I will leave them as is unless someone has strong feelings about it.